core/sr-primitives/src/curve.rs

```diff
@@ -0,0 +1,164 @@
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Provides some utilities to define a piecewise linear function.
+
+use crate::{Perbill, traits::{SimpleArithmetic, SaturatedConversion}};
+use core::ops::Sub;
+
+/// Piecewise Linear function in [0, 1] -> [0, 1].
+#[cfg_attr(feature = "std", derive(Debug))]
+#[derive(PartialEq, Eq)]
+pub struct PiecewiseLinear<'a> {
+	/// Array of points. Must be in order from the lowest abscissas to the highest.
+	pub points: &'a [(Perbill, Perbill)]
+}
+
+fn abs_sub<N: Ord + Sub<Output=N> + Clone>(a: N, b: N) -> N where {
+	a.clone().max(b.clone()) - a.min(b)
+}
+
+impl<'a> PiecewiseLinear<'a> {
+	/// Compute `f(n/d)*d` with `n <= d`. This is useful to avoid loss of precision.
+	pub fn calculate_for_fraction_times_denominator<N>(&self, n: N, d: N) -> N where
+		N: SimpleArithmetic + Clone
+	{
+		let n = n.min(d.clone());
+
+		if self.points.len() == 0 {
+			return N::zero()
+		}
+
+		let next_point_index = self.points.iter()
+			.position(|p| n < p.0 * d.clone());
+
+		let (prev, next) = if let Some(next_point_index) = next_point_index {
+			if let Some(previous_point_index) = next_point_index.checked_sub(1) {
+				(self.points[previous_point_index], self.points[next_point_index])
+			} else {
+				// There is no previous points, take first point ordinate
+				return self.points.first().map(|p| p.1).unwrap_or_else(Perbill::zero) * d
+			}
+		} else {
+			// There is no next points, take last point ordinate
+			return self.points.last().map(|p| p.1).unwrap_or_else(Perbill::zero) * d
+		};
+
+		let delta_y = multiply_by_rational_saturating(
+			abs_sub(n.clone(), prev.0 * d.clone()),
+			abs_sub(next.1.into_parts(), prev.1.into_parts()),
+			// Must not saturate as prev abscissa > next abscissa
+			next.0.into_parts().saturating_sub(prev.0.into_parts()),
+		);
+
+		// If both substration are same sign then result is positive
+		if (n > prev.0 * d.clone()) == (next.1.into_parts() > prev.1.into_parts()) {
+			(prev.1 * d).saturating_add(delta_y)
+		// Otherwise result is negative
+		} else {
+			(prev.1 * d).saturating_sub(delta_y)
+		}
+	}
+}
+
+// Compute value * p / q.
+// This is guarantee not to overflow on whatever values nor loose precision.
+// `q` must be superior to zero.
+fn multiply_by_rational_saturating<N>(value: N, p: u32, q: u32) -> N
```

Comment:
this stuff is okay for now but as me want to refactor more of the arithmatic stuff and move them into their own crate, I wonder: 

What is the accuracy requirement here? if p and q are always u32 and we want to be able to accurately compute it, then perquintillion is the option to go atm. 

On the other hand, that's a bit of on overkill and has some redundant bytes. If this pattern (`a * p / q` where `p` and `q` are `u32`) is common then actually making the `per-u32` might be worthwhile. It will also be a challenge to see if the macros in the sr-arithmatic refactor actually work or now 😀

Reply:
> What is the accuracy requirement here? if p and q are always u32 and we want to be able to accurately compute it, then perquintillion is the option to go atm.

well doesn't perquintillion::from_rational_approximation does loose a bit because like 1/3 is 33333.../quintillion but maybe we can afford this.
Or maybe now that I think of it we should go the other way around make perthings use this function 🤔

Reply:
I think what you mean is not what I meant by _accuracy_, but rather a _rounding error_ of at most `1` unit which, imo, we can afford. What I mean is: 

The most fine-grained approximation of this function is `p = 1, q = u32::MAX`. This will be roughly 1 divided by 4 billion. Naturally, per-1-billion is inherently incapable of representing this number while a per quintillion is. This is why to abstract this, something like per-billion is inaccurate, per quintillion is slightly wasteful while per-u32 is ideal.

Reply:
no what I meant is that there is an inherent imprecision from converting from one fraction to one another in some cases: let's say a = 100billion, p=1, q=3 converting 1/3 into a perbillion is inaccurate and dependings on how much a is bigger than a billion the inaccuracy remains:

```rust
	let a = 100 * 1_000_000_000u64;
	let res = multiply_by_rational_saturating(a, 1, 3);
	assert_eq!(res, 33333333333);
	let res = super::Perbill::from_rational_approximation(1u32, 3u32) * a;
	assert_eq!(res, 33333333300);
```